File: modules/time-zone/src/TimeZoneModule.ts
import { NativeModule, requireNativeModule } from 'expo';

declare class TimeZoneModule extends NativeModule {
  localTimeZoneIdentifier(): string;
  localTimeZoneDisplayName(): string;
  abbreviationDictionary(): Record<string, string>;
  abbreviationForIdentifier(identifier: string): string;
  identifierForAbbreviation(abbreviation: string): string;
  knownTimeZoneIdentifiers(): string[];
}

export default requireNativeModule<TimeZoneModule>('TimeZone');

File: modules/time-zone/ios/TimeZoneModule.swift
import ExpoModulesCore

public class TimeZoneModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TimeZone")

    Function("localTimeZoneIdentifier") {
      return TimeZone.current.identifier
    }

    Function("localTimeZoneDisplayName") {
      return TimeZone.current.localizedName(for: .standard, locale: Locale.current)
    }

    Function("abbreviationDictionary") {
      return TimeZone.abbreviationDictionary
    }

    Function("abbreviationForIdentifier") { (identifier: String) -> String? in
      return TimeZone(identifier: identifier)?.abbreviation()
    }

    Function("identifierForAbbreviation") { (abbreviation: String) -> String? in
      return TimeZone.abbreviationDictionary.first(where: { $0.value == abbreviation })?.key
    }

    Function("knownTimeZoneIdentifiers") {
      return TimeZone.knownTimeZoneIdentifiers
    }
  }
}
